src/main/java/bio/terra/service/filedata/DrsService.java

```diff
@@ -557,36 +561,82 @@ private DRSAccessURL signAzureUrl(
   }
 
   private DRSAccessURL signGoogleUrl(
-      SnapshotCacheResult cachedSnapshot, String gsPath, AuthenticatedUserRequest authUser) {
-    Storage storage;
+      SnapshotCacheResult cachedSnapshot,
+      String gsPath,
+      AuthenticatedUserRequest authUser,
+      String userProject) {
+    BlobId locator = GcsUriUtils.parseBlobUri(gsPath);
+
+    BlobInfo blobInfo = BlobInfo.newBuilder(locator).build();
+
+    String signingProject;
+    String signingUser;
+    // If a user specifies a billing project in the request, prefer that over the snapshot's
+    // project.  If a billing project is required to access the data, and it is not provided,
+    // nothing will fail until the user tries to access the signed URL.
+    if (!StringUtils.isEmpty(userProject)) {
+      signingProject = userProject;
+    } else {
+      signingProject = cachedSnapshot.googleProjectId;
+    }
+    if (authUser != null) {
+      signingUser = authUser.getEmail();
+    } else {
+      signingUser = null;
+    }
```

Comment:
This code only initializes `signingProject` and `signingUser` so they can be used by `getUrlSigningOptions()`. Since `getUrlSigningOptions()` is private and only called from this code, I would move this logic into that method and instead pass in the data the method needs to look at to compute the values. You end up passing in three args (`userProject`, `cachedSnapshot.googleProjectId` and `authUser`) but I think the resulting code is clearer.

Reply:
Yeah, that's a fair point.  I was going back and forth. I was going to try to use `getUrlSigningOptions` as a mockable method to be able to test that the right parameters were getting parsed out but then I ended up finding a (gross) way to mock the `Storage` objects